rpmlint/checks/SharedLibraryPolicyCheck.py

```diff
@@ -0,0 +1,174 @@
+        libname = libname.replace('.', '_')
+        return libname
+
+    def check(self, pkg):
+        if pkg.is_source:
+            return
+
+        # we don't care about develpkgs
+        if pkg.name.endswith('-devel'):
+            return
+
+        # check the pkg has any libname
+        libfiles = []
+        for fname in pkg.files.keys():
+            if not self.re_so_files.match(fname):
+                continue
+            libfiles.append(fname)
+
+        # if we didn't find any library files then we
+        # don't need to check anything
+        if not libfiles:
+            # verify if name does not match the slpp and if we still don't have any lib then error out
+            if self.re_soname_pkg.match(pkg.name):
+                self.output.add_info('E', pkg, 'shlib-policy-missing-lib')
+            return
+
+        # the soname validation matching the name is done
+        # already in BinaryCheck._check_shared_library
+        # Search for shared libraries in this package
+        libs = set()
+        libs_needed = set()
+        libs_to_dir = {}
+        dirs = set()
+        reqlibs = set()
+        pkg_requires = {x.name.split('(')[0] for x in pkg.requires}
+
+        for filename, pkgfile in pkg.files.items():
+            path = Path(filename)
+            if '.so.' in filename or filename.endswith('.so'):
+                if stat.S_ISREG(pkg.files[filename].mode) and pkgfile.magic.startswith('ELF '):
+                    readelf_parser = ReadelfParser(pkgfile.path, filename)
+                    failed_reason = readelf_parser.parsing_failed_reason()
+                    if failed_reason:
+                        self.output.add_info('E', pkg, 'readelf-failed', filename, failed_reason)
+                        return
+                    dyn_section = readelf_parser.dynamic_section_info
+                    libs_needed = libs_needed.union(dyn_section.needed)
+                    if dyn_section.soname:
+                        lib_dir = str(path.parent)
+                        libs.add(dyn_section.soname)
+                        libs_to_dir[dyn_section.soname] = lib_dir
+                        dirs.add(lib_dir)
+                    if dyn_section.soname in pkg_requires:
+                        # But not if the library is used by the pkg itself
+                        # This avoids program packages with their own
+                        # private lib
+                        # FIXME: we'd need to check if somebody else links
+                        # to this lib
+                        reqlibs.add(dyn_section.soname)
+
+        # If this is a program package (all libs it provides are
+        # required by itself), bail out
+        if not pkg.name.startswith('lib') and not libs.difference(reqlibs):
+            return
+
+        std_lib_package = False
+        pkg_ends_with_digit = pkg.name[-1].isdigit()
+        if pkg.name.startswith('lib') and pkg_ends_with_digit:
+            std_lib_package = True
```

Comment:
isn’t this checked by re_soname_pkg?

Reply:
I guess `re_soname_pkg` is more restrictive than `startswith('lib')`.

Reply:
We can likely use it..

Reply:
So we can't as the regex expect a trailing digit in package name.
And we need to catch: `E: shlib-policy-nonversioned-dir /usr/lib64/hello` in `libslp-missing-suffix.rpm` package.